### PR TITLE
`restart` lemmas in Tcb_R.thy

### PR DESCRIPTION
**Pull request by MitchellBuckley**

### CODE REVIEW

File: proof/refine/ARM/Tcb_R.thy [hunk 1]
```diff
@@ -294,59 +294,130 @@ lemma setCTE_weak_sch_act_wf[wp]:
...
+            apply (case_tac scOpt; clarsimp)
+             apply (fastforce)
+            apply (subst (asm) sym_refs_bound_sc_tcb_iff_sc_tcb_sc_at[OF eq_commute refl])
+             apply (fastforce)
+            apply (fastforce simp: sc_at_pred_n_def obj_at_def sym_refs_bound_sc_tcb_iff_sc_tcb_sc_at[OF eq_commute refl])
+           apply (wpsimp wp: setThreadState_Restart_invs')
+          apply clarsimp
+          apply (rule_tac Q="\<lambda>rv. invs and valid_sched_action and tcb_at t
+          and fault_tcb_at ((=) None) t and st_tcb_at (\<lambda>st'. tcb_st_refs_of st' = {}) t
+     and scheduler_act_not t and bound_sc_tcb_at ((=) scOpt) t
```
Comment (MitchellBuckley) [inline proof/refine/ARM/Tcb_R.thy:372]: indentation

File: proof/refine/ARM/Tcb_R.thy [hunk 2]
```diff
@@ -294,59 +294,130 @@ lemma setCTE_weak_sch_act_wf[wp]:
...
+  apply (fastforce dest: global'_no_ex_cap)
+  done
+
+crunches cancel_ipc
+  for valid_sched_action[wp]: valid_sched_action
+  (wp: crunch_wps ignore: set_object thread_set update_sched_context)
+
 lemma restart_corres:
-  "corres dc (einvs  and tcb_at t) (invs' and tcb_at' t)
+  "corres dc (einvs  and tcb_at t and ex_nonz_cap_to t) (invs' and tcb_at' t and ex_nonz_cap_to' t)
```
Comment (VictorPhan) [inline proof/refine/ARM/Tcb_R.thy:333]: nitpick: remove extra space after `einvs`

File: proof/refine/ARM/Tcb_R.thy [hunk 3]
```diff
@@ -294,59 +294,130 @@ lemma setCTE_weak_sch_act_wf[wp]:
...
+                  apply (wpsimp wp: is_schedulable_wp isSchedulable_wp)+
+               apply (rule_tac Q="\<lambda>rv. invs and valid_sched_action and tcb_at t" in hoare_strengthen_post)
+                apply (wpsimp wp: sched_context_resume_valid_sched_action)
+               apply fastforce
+              apply (rule_tac Q="\<lambda>rv. invs' and tcb_at' t" in hoare_strengthen_post)
+               apply wpsimp
+              apply (fastforce simp: invs'_def valid_state'_def sch_act_wf_weak valid_pspace'_def)
+             apply clarsimp
+            apply (rule_tac Q="\<lambda>rv. invs and valid_sched_action and tcb_at t
+                                    and (scheduler_act_not t)
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Tcb_R.thy:360]: I think these brackets are unnecessary.

Also some extra brackets around `fastforce`s below

**Review by michaelmcinerney — APPROVED** [review]
Lovely

**Review by corlewis — APPROVED** [review]

**Review by mktnk3 — APPROVED** [review]
Nice :)

**Pull request merged**